### PR TITLE
report the ddrace team to master instead of the normal team

### PR DESCRIPTION
**Pull request by dobrykafe**
## Checklist

- [ ] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by heinrich5991** [comment]
Maybe we still want to use `-1` for spectators. Other than that, looks good to me.

**Comment by dobrykafe** [comment]
> Maybe we still want to use `-1` for spectators. Other than that, looks good to me.

ok done

**Comment by ChillerDragon** [comment]
``TEAM_SPECTATORS`` would be nicer than the magic value ``-1`` also rip downstream mods that have team based gameplay :(

I guess as of right now that team is not used anywhere anyways right? Just in some community stats projects.

**Comment by ChillerDragon** [comment]
Would be epic if it still uses regular teams if the active controller has the ``GAMEFLAG_TEAMS`` set

File: src/game/server/gamecontext.cpp [hunk 1]
```diff
@@ -4653,5 +4653,5 @@ void CGameContext::OnUpdatePlayerServerInfo(char *aBuf, int BufSize, int ID)
 		"\"team\":%d",
 		aJsonSkin,
 		JsonBool(m_apPlayers[ID]->IsAfk()),
-		m_apPlayers[ID]->GetTeam());
+		m_apPlayers[ID]->GetTeam() == -1 ? -1 : GetDDRaceTeam(ID));
```
Comment (heinrich5991) [inline src/game/server/gamecontext.cpp:4656]:
```suggestion
		m_apPlayers[ID]->GetTeam() == TEAM_SPECTATORS ? -1 : GetDDRaceTeam(ID));
```

The other -1 shouldn't be replaced.

**Comment by heinrich5991** [comment]
> I guess as of right now that team is not used anywhere anyways right?

Correct.

> Just in some community stats projects.

Is it?

**Comment by ChillerDragon** [comment]
> > Just in some community stats projects.
> 
> 
> 
> Is it?

I assumed ryo tracks and graphs the entire json. My point was more that even if the ddnet client is not using it someone might already be looking at those values with some kind of tool.

File: src/game/server/gamecontext.cpp (hunk 1 above)
Comment (ChillerDragon) [reply src/game/server/gamecontext.cpp:4656]: Interesting 🤔

**Comment by dobrykafe** [comment]
> Would be epic if it still uses regular teams if the active controller has the `GAMEFLAG_TEAMS` set

tried implementing this. not sure if its correct

**Comment by ChillerDragon** [comment]
> > Would be epic if it still uses regular teams if the active controller has the `GAMEFLAG_TEAMS` set
> 
> 
> 
> tried implementing this. not sure if its correct

Cool thanks yea that is what I meant.

**Comment by ChillerDragon** [comment]
Double tenary 🐼

**Comment by heinrich5991** [comment]
Given that this now includes vanilla teams as well, we need to think about the interactions. Do gamemodes include vanilla teams and DDNet teams at the same time? What should happen in this case?

File: src/game/server/player.cpp [hunk 2]
```diff
@@ -626,6 +626,8 @@ void CPlayer::SetTeam(int Team, bool DoChatMsg)
 				pPlayer->m_SpectatorID = SPEC_FREEVIEW;
 		}
 	}
+
+	Server()->ExpireServerInfo();
```
Comment (heinrich5991) [inline src/game/server/player.cpp:630]:
```suggestion
	if(!GameServer()->m_pController->IsTeamPlay())
	{
		Server()->ExpireServerInfo();
	}
```

**Review by heinrich5991 — COMMENTED** [review]
LGTM otherwise. :)

**Comment by edg-l** [comment]
actually this was my intention when making it, i guess i got it wrong and sent the vanilla team (?9

**Review by heinrich5991 — APPROVED** [review]
Thanks!

**Comment by ChillerDragon** [comment]
> actually this was my intention when making it, i guess i got it wrong and sent the vanilla team (?)

Code base is the trollest. I also never know which getter returns what team hehe

File: src/game/server/player.cpp [hunk 3]
```diff
@@ -626,6 +626,11 @@ void CPlayer::SetTeam(int Team, bool DoChatMsg)
 				pPlayer->m_SpectatorID = SPEC_FREEVIEW;
 		}
 	}
+
+	if(!GameServer()->m_pController->IsTeamPlay())
```
Comment (ChillerDragon) [inline src/game/server/player.cpp:630]: They way I understand it this does not update on join spec/red/blue in master server when the teams are on. What is the reason for that?

https://github.com/ZillyInsta/ddnet-insta/issues/85

Comment (ChillerDragon) [reply src/game/server/player.cpp:630]: https://github.com/ddnet/ddnet/pull/7794

Comment (ChillerDragon) [reply src/game/server/player.cpp:630]: Now that I think about it you probably want another expire server info in teams.cpp for example in

``void CGameTeams::SetForceCharacterTeam(int ClientID, int Team)``

To update the master when the ddrace team changed

**Pull request merged**